rfd/0078-role-variables.md

```diff
@@ -0,0 +1,176 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 78 - Role Variables
+
+## What
+
+Teleport admins should be able to define variables within roles which can extend
+claims coming from internal and external user traits.
+
+Requirements:
+
+1. Variable definitions should support some form of pattern matching so that
+   traits can be easily and powerfully extended.
+2. A variable defined in one role should be usable in many other roles.
+
+Non-requirements:
+
+1. Variables can depend on other variables.
+
+## Why
+
+Variables will be useful to specify additional traits that are not available in
+OIDC/SAML claims.
+
+Teleport admins don't always have control over their org's OIDC/SAML provider.
+
+Role variables will simplify role evaluation, keep logic in one place, and
+minimize repetition.
+
+## Details
+
+Example role spec:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: node-users
+spec:
+  vars:
+    - name: "logins"
+      # "select" is a string value which will be used as the input to all
+      # possible expansion matchers. It can (and usually will) contain a trait
+      # which will be expanded before passing as input.
+      select: "{{external.email}}"
+      expansions:
+        # nic should be able to login as root
+        - match: '^nic@goteleport\.com$'
```

Comment:
I wonder if we are leaning too heavily on regular expressions here. While powerful, they have limitations in expressiveness and are hard to read; this makes role security auditing harder than it should be. Perhaps we can use expression language we support for traits, with additional extensions as needed?

Reply:
What do you think this would look like with an expression language?

Reply:
There are various ways to cut it, but we should aim for a clean expression of common patterns. So, what common use cases would we like to see?

Reply:
I have to say, I think regular expressions are a pretty powerful solution here. Most people are pretty familiar with them.

I can't even really think of what this would like like with an expression language, but I think you would mostly just want to do regex-like matching in that language. If you have a counter example I'd be interested to see it.

One "killer feature" of regex's here, that perhaps I haven't highlighted enough, is the ability to use capture groups to use parts of the input in the output value, like below. I think this is basically a necessary feature here and I'm not sure of a better way to enable it than regular expressions.
```
- match: '^env-(\w+)$'
  out: '$1'
```

Reply:
An expression language is a big thing to design and implement, so perhaps regexes offer a safe path forwards. Nevertheless, I'd consider using something like CEL (https://github.com/google/cel-spec, https://github.com/google/cel-go).

Reply:
I've added examples under the alternatives section written with the `predicate` language as well as CEL.

It's an interesting solution and it's pretty fun to experiment with, but it does feel complex and I can imagine fighting with it as a user. Not sure if what I came up with is similar to what you were thinking.